src/palette.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+type Palette map[string]string
+
+const (
+	paletteKeyError          = "palette: requested color %s does not exist in palette of colors %s"
+	paletteRecursiveKeyError = "palette: resolution of color %s returned palette reference %s; recursive references are not supported"
+)
+
+var (
+	paletteColorPrefixes = [...]string{"palette:", "p:"}
+)
+
+// ResolveColor gets a color value from the palette using given colorName.
+// If colorName is not a palette reference, it is returned as is.
+func (p Palette) ResolveColor(colorName string) (string, error) {
+	key, ok := p.asPaletteKey(colorName)
+	// colorName is not a palette key, return it as is
+	if !ok {
+		return colorName, nil
+	}
+
+	color, ok := p[key]
+	if !ok {
+		return "", &PaletteKeyError{Key: key, palette: p}
+	}
+
+	if _, ok = p.asPaletteKey(color); ok {
+		return "", &PaletteRecursiveKeyError{Key: key, Value: color}
```

Comment:
this is actually solvable by looping the above as long as you find the prefix and a match. In my sample I wrote the following:

```golang
getPaletteColor := func(color string) string {
    mapPaletteColor := func(color string) string {
        color = strings.TrimLeft(color, "p:")
        if color, found := a.palette[color]; found {
		    return color
	    }
	    return "#111111"
    }
    for strings.HasPrefix(color, "p:") {
	    color = mapPaletteColor(color)
    }
    return color
}
```

I know you could argue why, but call this experience, this is going to happen anyways. So we better get prepared if possible ;-)

Reply:
This will run into infinite recursion in case of, for example, `"palette": {"blue": "p:blue"}`. I added the recursion (with depth check), but that actually made the code more complicated than it can be.

Reply:
And, in the case of `for` loop, infinite recursion will never terminate.

Reply:
that's solvable by adding 1 check:

```golang
getPaletteColor := func(color string) string {
    mapPaletteColor := func(color string) string {
        color = strings.TrimLeft(color, "p:")
        if color, found := a.palette[color]; found {
		    return color
	    }
	    return "#111111"
    }
    for strings.HasPrefix(color, "p:") {
	    refColor = mapPaletteColor(color)
            if refColor == color {
                break
            }
    return color
}
```

Reply:
You're right. Did some tests, there's no simple around this without caching all entries in the path. I would set the depth to 3 through. No need to go crazy here.